### PR TITLE
enh(clapi): handle host/hosttemplate macros write in vault with CLAPI

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to handle the writing of Host & Host Templates Macros in vault Host & Host Templates through CLAPI.
This PR handle the following actions:
- set a macro
- delete a macro
- export the macro (the raw value will be retrieve from the vault before added to the export file)

**Fixes** # MON-16393

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [x] 23.04.x (master)

<h2> How this pull request can be tested ? </h2>

- Add a vault configuration to your platform
- Create an host
- Create a macro as password for your host using clapi
- Your macro should be store in vault and the path should be stored in centreon.on_demand_macro_host table
- Export your configuration with CLAPI
- The raw value of your macro should appears 
- Delete your macro using clapi
- The macro should be delete from the vault, if it was the only macro stored on this endpoint, the endpoint should be delete.

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/www/class/centreon-clapi/centreonHost.class.php [hunk 1]
```diff
@@ -1510,6 +1590,28 @@ public function export($filterName = null)
                     $description = "'" . $description . "'";
                 }
 
+                // Retrieve raw value of the macro if it is recorded in vault.
+                if (
+                    $macro['is_password'] === 1
+                    && preg_match(self::VAULT_PATH_REGEX, $macro['host_macro_value'])
+                    && $this->getVaultConfiguration() !== null
```
Comment (jeremyjaouen) [inline centreon/www/class/centreon-clapi/centreonHost.class.php:1597]:
```suggestion
                    && $vaultConfiguration() !== null
```

File: centreon/www/class/centreon-clapi/centreonHost.class.php [hunk 2]
```diff
@@ -1010,6 +1013,49 @@ public function setmacro($parameters)
                 )
             );
         }
+        // If the macro is password and a vault configuration exists
+        if ((int) $params[3] === 1 && ($vaultConfiguration = $this->getVaultConfiguration()) !== null) {
```
Comment (jeremyjaouen) [inline centreon/www/class/centreon-clapi/centreonHost.class.php:1017]: rename $params[index] by evocating name

File: centreon/www/class/centreon-clapi/centreonHost.class.php [hunk 3]
```diff
@@ -1010,6 +1013,49 @@ public function setmacro($parameters)
...
+            );
+            $hostSecrets['_HOST' . strtoupper($params[1])] = $params[2];
+            $this->writeSecretsInVault(
+                $vaultConfiguration,
+                $hostId,
+                $clientToken,
+                $hostSecrets,
+                $logger,
+                $httpClient
+            );
```
Comment (jeremyjaouen) [inline centreon/www/class/centreon-clapi/centreonHost.class.php:1037]:
```suggestion
            $logger = $this->getLogger();
            $httpClient = new \CentreonRestHttp();
            $clientToken = $this->authenticateToVault($vaultConfiguration, $logger, $httpClient);
            $resourceEndpoint = $vaultConfiguration->getStorage() . '/monitoring/hosts/' . $hostId;
            $hostSecrets = $this->getHostSecretsFromVault(
                $vaultConfiguration,
                $resourceEndpoint,
                $clientToken,
                $logger,
                $httpClient
            );
            $hostSecrets['_HOST' . strtoupper($params[1])] = $params[2];
            $this->writeSecretsInVault(
                $vaultConfiguration,
                $hostId,
                $clientToken,
                $hostSecrets,
                $logger,
                $httpClient
            );
```
refactor into a method writeMacroSecretsInVault()

File: centreon/www/class/centreon-clapi/centreonHost.class.php [hunk 4]
```diff
@@ -1010,6 +1013,49 @@ public function setmacro($parameters)
...
+                $statement = $this->db->prepare(
+                    "SELECT host_macro_id FROM on_demand_macro_host WHERE host_macro_name = :macroName AND host_host_id = :hostId"
+                );
+                $statement->bindValue(':macroName', $this->wrapMacro($params[1]), \PDO::PARAM_STR);
+                $statement->bindValue(':hostId', $hostId, \PDO::PARAM_INT);
+                $statement->execute();
+                if (($result = $statement->fetch()) !== false) {
+                    $macroId = $result['host_macro_id'];
+                }
+            }
```
Comment (jeremyjaouen) [inline centreon/www/class/centreon-clapi/centreonHost.class.php:1050]: get macroId from return of insert

File: centreon/www/class/centreon-clapi/centreonHost.class.php [hunk 5]
```diff
@@ -1510,6 +1590,28 @@ public function export($filterName = null)
...
+                        $hostSecrets = $this->getHostSecretsFromVault(
+                            $vaultConfiguration,
+                            $resourceEndpoint,
+                            $clientToken,
+                            $logger,
+                            $httpClient
+                        );
+                        if(array_key_exists(trim($macro['host_macro_name'], "$"), $hostSecrets) ) {
+                            $macro['host_macro_value'] = $hostSecrets[trim($macro['host_macro_name'], "$")];
+                        }
```
Comment (jeremyjaouen) [inline centreon/www/class/centreon-clapi/centreonHost.class.php:1612]: indentation

Comment (jeremyjaouen) [reply centreon/www/class/centreon-clapi/centreonHost.class.php:1612]: check hashicorp API doc to bulk requests

File: centreon/www/class/centreon-clapi/centreonHost.class.php [hunk 6]
```diff
@@ -1510,6 +1590,28 @@ public function export($filterName = null)
...
+                // Retrieve raw value of the macro if it is recorded in vault.
+                if (
+                    $macro['is_password'] === 1
+                    && preg_match(self::VAULT_PATH_REGEX, $macro['host_macro_value'])
+                    && $vaultConfiguration() !== null
+                ) {
+                    $logger = $this->getLogger();
+                        $httpClient = new \CentreonRestHttp();
+                        $clientToken = $this->authenticateToVault($vaultConfiguration, $logger, $httpClient);
+                        $resourceEndpoint = preg_replace(self::VAULT_PATH_REGEX, "", $macro['host_macro_value']);
```
Comment (jeremyjaouen) [inline centreon/www/class/centreon-clapi/centreonHost.class.php:1602]: authenticate before foreach

**Review by wtermellil — APPROVED** [review]

**Comment by jeremyjaouen** [comment]
this will be handled later

**Pull request closed without merging**